File: src/ui/packages/client/src/shellCapture.ts
// Hand-written TypeScript mirror of the shell capture domain DTO in
// `src/domain/src/shell_capture.rs`. There is no TypeScript codegen; this
// interface tracks the Rust type field-for-field using serde's snake_case
// field names. Callers run a command with
// `client.call('shell.run_capture', { command })`, which resolves to a
// `ShellCaptureResult`.

/**
 * The captured result of running a launcher command: its command line, the
 * standard output and standard error it produced, the process exit code, and
 * whether it was terminated because it exceeded the allotted time. Mirrors the
 * Rust `ShellCaptureResult`.
 */
export interface ShellCaptureResult {
  command: string;
  stdout: string;
  stderr: string;
  exit_code: number;
  timed_out: boolean;
}

File: src/domain/src/shell_capture.rs
use serde::{Deserialize, Serialize};

/// The captured result of running a launcher command: its command line, the
/// standard output and standard error it produced, the process exit code, and
/// whether it was terminated because it exceeded the allotted time.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ShellCaptureResult {
    pub command: String,
    pub stdout: String,
    pub stderr: String,
    pub exit_code: i32,
    pub timed_out: bool,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn shell_capture_result_json_round_trips() {
        let result = ShellCaptureResult {
            command: "echo hello".to_string(),
            stdout: "hello\n".to_string(),
            stderr: String::new(),
            exit_code: 0,
            timed_out: false,
        };
        let json = serde_json::to_string(&result).expect("serialize");
        let parsed: ShellCaptureResult = serde_json::from_str(&json).expect("deserialize");
        assert_eq!(parsed, result);
    }
}
